lib/unleash/repo.ex

```diff
@@ -1,124 +1,128 @@
 
   def start_link(state) do
     {:ok, pid} = GenServer.start_link(__MODULE__, state, name: Unleash.Repo)
 
-    unless Config.test? do
+    unless Config.test?() do
       initialize()
     end
 
     {:ok, pid}
   end
 
   def get_feature(name) do
-    GenServer.call(Unleash.Repo, {:get_feature, name}, 100)
+    Cache.get_feature(name)
   end
 
   def get_all_feature_names do
-    GenServer.call(Unleash.Repo, {:get_all_feature_names}, 100)
-  end
-
-  def handle_call({:get_feature, name}, _from, state) do
-    feature = Features.get_feature(state, name)
-
-    {:reply, feature, state}
-  end
-
-  def handle_call({:get_all_feature_names}, _from, state) do
-    names = Features.get_all_feature_names(state)
-    {:reply, names, state}
+    Cache.get_all_feature_names()
   end
 
-  def handle_info({:initialize, etag, retries}, state) do
+  def handle_info({:initialize, etag, retries}, _state) do
     if retries > 0 or retries <= -1 do
+      cached_features = %Features{features: Cache.get_features()}
+
       {etag, response} =
         case Unleash.Config.client().features(etag) do
-          :cached -> {etag, state}
+          :cached -> {etag, cached_features}
           x -> x
         end
 
-      features =
+      remote_features =
         case response do
           {:error, _} ->
-            state
-            |> read_state()
+            cached_features
+            |> read_file_state()
             |> schedule_features(etag, retries - 1)
 
           f ->
             schedule_features(f, etag)
         end
 
-      if features === state do
-        {:noreply, state}
+      if remote_features === cached_features do
+        {:noreply, nil}
       else
-        write_state(features)
+        Cache.upsert_features(remote_features.features)
+        write_file_state(remote_features)
+        {:noreply, nil}
       end
     else
       Logger.debug(fn ->
         "Retries === 0, disabling polling"
       end)
 
-      {:noreply, state}
+      {:noreply, nil}
     end
   end
 
   # https://github.com/appcues/mojito/issues/57
   # Work around for messages received from Mojito after we've passed over the timeout
   # threshold.
-  def handle_info({:mojito_response, _ref, _message}, state) do
-    {:noreply, state}
+  def handle_info({:mojito_response, _ref, _message}, _state) do
+    {:noreply, nil}
   end
 
-  defp read_state(%Features{features: []} = state) do
+  defp read_file_state(%Features{features: []} = cached_features) do
     if File.exists?(Config.backup_file()) do
       Config.backup_file()
       |> File.read!()
       |> Jason.decode!()
       |> Features.from_map!()
     else
-      state
+      cached_features
     end
   end
 
-  defp read_state(state), do: state
+  defp read_file_state(cached_features), do: cached_features
 
-  defp write_state(state) do
+  defp write_file_state(features) do
     if not File.dir?(Config.backup_dir()) do
       Config.backup_dir()
-      |> File.mkdir_p!()
+      |> File.mkdir_p()
     end
```

Comment:
what failure behavior do you want here? With this change any failures will be ignored, but it will instead fail on the `File.write! a few lines down, and crash the same

Reply:
As per [Sasha's comment](https://github.com/duffelhq/unleashed_ex/pull/2#discussion_r752141253) this change was to blindly create the whole directory path rather than check for it first (although i missed that in my code)

I've added sasha's full suggestion, it should raise a matcherror upon failure